zebra/zebra_evpn_arp_nd.c

```diff
@@ -0,0 +1,651 @@
+	uint16_t vlan = 0;
+	struct zebra_if *zif;
+	int fd;
+	int len;
+	uint8_t buf[ZEBRA_EVPN_ARP_ND_MAX_PKT_LEN];
+
+	zif = THREAD_ARG(t);
+	fd = THREAD_FD(t);
+
+	for (count = 0; count < ZEBRA_EVPN_ARP_ND_PKT_MAX; ++count) {
+		if (zebra_evpn_arp_nd_recvmsg(fd, buf, sizeof(buf), &vlan, &len)
+		    < 0) {
+			if (errno == EINTR)
+				continue;
+
+			if (errno == EWOULDBLOCK || errno == EAGAIN)
+				break;
+
+			if (IS_ZEBRA_DEBUG_EVPN_MH_ARP_ND_PKT)
+				zlog_debug("evpn arp_nd read failed:len %d %s",
+					   len, safe_strerror(errno));
+			break;
+		}
+
+		if (zebra_evpn_arp_nd_proc(zif, vlan, buf, len) < 0)
+			break;
+	}
+
+	/* prepare for next installment of packets */
+	zebra_evpn_arp_nd_pkt_read_enable(zif);
+
+	return 0;
+}
+
+/* BPF filter for snooping on ARP replies and IPv6 Neighbor advertisements -
+ * tcpdump -dd '((arp and arp[6:2] == 2)
+ *			or (icmp6 and ip6[40] == 136)) and inbound'
+ */
+static struct sock_filter arp_nd_reply_filter[] = {
+	{0x28, 0, 0, 0x0000000c},  {0x15, 0, 2, 0x00000806},
+	{0x28, 0, 0, 0x00000014},  {0x15, 8, 11, 0x00000002},
+	{0x15, 0, 10, 0x000086dd}, {0x30, 0, 0, 0x00000014},
+	{0x15, 3, 0, 0x0000003a},  {0x15, 0, 7, 0x0000002c},
+	{0x30, 0, 0, 0x00000036},  {0x15, 0, 5, 0x0000003a},
+	{0x30, 0, 0, 0x00000036},  {0x15, 0, 3, 0x00000088},
+	{0x28, 0, 0, 0xfffff004},  {0x15, 1, 0, 0x00000004},
+	{0x6, 0, 0, 0x00040000},   {0x6, 0, 0, 0x00000000},
+};
+
+/* Setup socket per-access bridge port */
+static int zebra_evpn_arp_nd_sock_create(struct zebra_if *zif)
+{
+	int fd;
+	int reuse = 1;
+	int rcvbuf = ZEBRA_EVPN_ARP_ND_SOC_RCVBUF;
+	long flags;
+	struct sock_fprog prog = {
+		.len = sizeof(arp_nd_reply_filter)
+		       / sizeof(arp_nd_reply_filter[0]),
+		.filter = arp_nd_reply_filter,
+	};
+
+	frr_with_privs (&zserv_privs) {
+		fd = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+	}
+
+	if (fd < 0) {
+		flog_err(EC_LIB_SOCKET, "evpn arp_nd sock create: errno %s",
+			 safe_strerror(errno));
+		return -1;
+	}
+
+	if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (void *)&reuse,
+		       sizeof(reuse))) {
+		flog_err(EC_LIB_SOCKET,
+			 "evpn arp_nd sock SO_REUSEADDR set: fd %d errno %s",
+			 fd, safe_strerror(errno));
+		close(fd);
+		return -1;
+	}
+
+	frr_with_privs (&zserv_privs) {
+		if (setsockopt(fd, SOL_SOCKET, SO_RCVBUFFORCE, &rcvbuf,
+			       sizeof(rcvbuf)))
+			zlog_warn(
+				"evpn arp_nd sock Failure to set rcvbuf to %d errno %s",
+				rcvbuf, safe_strerror(errno));
+	}
+
+	/* enable aux data for getting the vlan if (kernel strips the
+	 * vlan tag for raw sockets)
+	 */
+	if (setsockopt(fd, SOL_PACKET, PACKET_AUXDATA, (void *)&reuse,
+		       sizeof(reuse))) {
+		flog_err(EC_LIB_SOCKET,
+			 "evpn arp_nd sock PACKET_AUXDATA set: fd %d errno %s",
+			 fd, safe_strerror(errno));
+		close(fd);
+		return -1;
+	}
+
+	if (setsockopt(fd, SOL_SOCKET, SO_ATTACH_FILTER, &prog, sizeof(prog))
+	    < 0) {
+		flog_err(EC_LIB_SOCKET,
+			 "evpn arp_nd sock PACKET_AUXDATA set: fd %d errno %s",
```

Comment:
error says PACKET_AUXDATA instead of SO_ATTACH_FILTER

Reply:
ack, fixed up